### PR TITLE
Value-Set Arithmetic

### PR DESCRIPTION
**Pull request by ZhangZhuoSJTU**
This PR is used to track my work on math operations of VSA, which including two stages:

- [x] Strided-Interval Arithmetic
- [x] Value-Set Arithmetic

More discussion could be found in #121

### CODE REVIEW

**Comment by ZhangZhuoSJTU** [comment]
Right now, `StridedInterval` has supported four fundamental operations. I will support bitwise operations as soon as possible.

Compared with original `analysis/valueset/strided_interval.rs`, new `StridedInterval` have much more features.

+ **Support K-bits Strided Interval**: The original one only consider 64-bits operations. However, a sound value set should not only support more than 64-bits, but at least 32-bits and 16-bits, saying `eax` and `ax`. In the new implement, we could support operations on any bits (0< K <= 64)

+ **Much More Correctness**: The correctness of original one cannot be guaranteed. Looking at following code in original one:
```  rust
impl Add for StridedInterval_n {
    type Output = StridedInterval_n;

    fn add(self, other: StridedInterval_n) -> StridedInterval_n {
        let (gcd, lcm) = gcd_lcm(self.stride, other.stride);
        StridedInterval_n::from(gcd,
                                self.min + other.min,
                                self.max + other.max,)
    }
}
```
If the addition between `self.min` and `other.min` is a regular one, but `self.max` plus `other.max` will cause overflow, the result should be `1[i64::MIN, i64::MAX]`, because the steps cannot still be a gcd any more.

New code do a lot concern on this:
```rust
impl Add for StridedInterval {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        if self.k != other.k {
            radeco_err!("Addition between two strided intervals with different radices");
            StridedInterval::default()
        } else if (self.constant().is_some()) && (other.constant().is_some()) {
            // two constants
            let n: inum = n_in_k_bits!(self.lb.wrapping_add(other.lb), self.k);
            StridedInterval::from((self.k, n))
        } else {
            let lb: inum = n_in_k_bits!(self.lb.wrapping_add(other.lb), self.k);
            let ub: inum = n_in_k_bits!(self.ub.wrapping_add(other.ub), self.k);
            // u < 0 indicates:
            //      self.lb + other.lb < min_value &&
            //          self.ub + other.ub > min_value (on k-bits space)
            let u: inum = n_in_k_bits!(self.lb & other.lb & !lb &
                                       !(self.ub & other.ub & !ub), self.k);
            // v < 0 indicates:
            //      min_value <= self.lb + other.lb <= max_value &&
            //          self.ub + other.ub > max_value (on k-bits space)
            let v: inum = n_in_k_bits!(((self.lb ^ other.lb) | !(self.lb ^ lb)) &
                                       (!self.ub & !other.ub & ub), self.k);
            if (u < 0) || (v < 0) {
                StridedInterval::default_k(self.k)
            } else {
                StridedInterval::new(
                    self.k,
                    gcd(self.s, other.s),
                    lb,
                    ub
                )
            }
        }
    }
}
```

There are many other places in original code which leads to wrong results.

+ **Much More Precision**: The original one care less about special cases, the new one do a lot, to make a huge increase on precision of operation.  Like when a value set is multiplied by zero, the new code will generate a constant of zero. 

+ **More Tests**: There is enough tests for new code to make sure the logic is correct.


@XVilka @sushant94 @chinmaydd If you could give me some suggestions about this part of code, I will be very appreciated. I am trying to make sure the imprecision generated by Strided Interval as small as possible.

File: src/analysis/vsa/abstract_set/BDD.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2015, The Radare Project. All rights reserved.
```
Comment (XVilka) [inline src/analysis/vsa/abstract_set/BDD.rs]: Update copyrights then too

**Comment by XVilka** [comment]
So far looks good.

**Comment by XVilka** [comment]
@sushant94 so, what do you think?

**Comment by ZhangZhuoSJTU** [comment]
The bitwies is a little bit complex, I am trying my best to make test pass.

**Comment by XVilka** [comment]
@sushant94 @chinmaydd @ZhangZhuoSJTU ping

**Comment by ZhangZhuoSJTU** [comment]
There is only few functions to be done. I will do it as soon as possible. Thus, this PR will be closed after at most 24 hours. Please check if you are free ;). @XVilka @sushant94 

After this PR is done, I will try to go from flag operations back to the original comparisons for x86/x64. Then intra-procedural VSA will be possible.

**Comment by sushant94** [comment]
I have reviewed the overall design of this module, not if the algorithm is correct/bug free (this is too complicated to do and we should likely do through tests). I think the current tests are sufficient and docs are ok. SGTM to merge, please do whenever @ZhangZhuoSJTU :)

**Comment by ZhangZhuoSJTU** [comment]
Merged.

**Pull request merged**